### PR TITLE
TypeScript module resolution using Metro's Graph

### PR DESCRIPTION
**Pull request by afoxman**
### Description

Add a custom TypeScript resolver which is built on top of Metro's Graph. TypeScript asks the resolver to map from a module to a source file (`.ts`, `.tsx`) or declaration file (`.d.ts`), which it will then use for type-checking. 

This PR implements a custom TypeScript resolver that operates on data from Metro's Graph (the output of its own resolver). The graph already incorporates platform-overrides (`foo.ios.ts`) and module name substitutions (`react-native` --> `react-native-windows`).
 
This is specifically for making Metro + TypeScript work together in the context of bundling and serving.

### Test plan

Manual bundle testing use the test-app package.

Run bundling with TypeScript's stock resolver and with the custom resolver. Compare each module -> resolved-file decision and make sure they align, except where platform-overrides and module-name-substitutions occur.

### CODE REVIEW

File: packages/typescript-service/src/resolve.ts [hunk 1]
```diff
@@ -1,90 +1,112 @@
...
-    resolveModuleNames,
-    getResolvedModuleWithFailedLookupLocationsFromCache,
-    resolveTypeReferenceDirectives,
+    resolveTypeReferenceDirectives: (
+      typeDirectiveNames: string[],
+      containingFile: string,
+      redirectedReference?: ts.ResolvedProjectReference
+    ): (ts.ResolvedTypeReferenceDirective | undefined)[] => {
+      const resolved: (ts.ResolvedTypeReferenceDirective | undefined)[] = [];
+      for (const name of typeDirectiveNames) {
```
Comment (tido64) [inline packages/typescript-service/src/resolve.ts:99]: Nit: If we use a map here, we shouldn't need to type up `resolved`.

Comment (tido64) [reply packages/typescript-service/src/resolve.ts:99]: Same goes for other instances of for-of.

File: packages/cli/src/metro-config.ts [hunk 2]
```diff
@@ -6,19 +6,87 @@ import {
   DuplicateDependencies,
   Options as DuplicateDetectorOptions,
 } from "@rnx-kit/metro-plugin-duplicates-checker";
-import type { Project } from "@rnx-kit/typescript-service";
 import { MetroPlugin, MetroSerializer } from "@rnx-kit/metro-serializer";
 import { MetroSerializer as MetroSerializerEsbuild } from "@rnx-kit/metro-serializer-esbuild";
 import type { AllPlatforms } from "@rnx-kit/tools-react-native/platform";
+import { Project } from "@rnx-kit/typescript-service";
```
[the remark this answers is not in the document]
Comment (afoxman) [reply packages/cli/src/metro-config.ts:12]: fixed

File: packages/test-app/src/App.native.tsx [hunk 3]
```diff
@@ -1,11 +1,3 @@
-/**
- * Sample React Native App
- * https://github.com/facebook/react-native
- *
- * @format
- * @flow strict-local
- */
-
 import React from "react";
```
[the remark this answers is not in the document]
Comment (afoxman) [reply packages/test-app/src/App.native.tsx:1]: I wanted it to have a platform override in the name to ensure that the new resolvers would pick it up, and fail if they didn't.

File: packages/cli/src/metro-ts-resolver.ts [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+
+  public resolveModuleNames(
+    moduleNames: string[],
+    containingFile: string,
+    _reusedNames: string[] | undefined,
+    _redirectedReference?: ResolvedProjectReference
+  ): (ResolvedModuleFull | undefined)[] {
+    let resolved: (string | undefined)[] | undefined = undefined;
+
+    console.log(`${containingFile}`);
```
[the remark this answers is not in the document]
Comment (afoxman) [reply packages/cli/src/metro-ts-resolver.ts:354]: I meant to remove those when I upgraded from draft to real review. They're removed now.

**Review by tido64 — APPROVED** [review]

**Pull request merged**